{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TRUST4 Exercise\n",
    "\n",
    "This notebook contains the implementation and descriptions of the TRUST4 excercise from Numenos.\n",
    "\n",
    "## Introduction\n",
    "The notebook contains the following main sections:\n",
    "1. `Configurations` - installs and generates reference data, and installs the TRUST4 docker image. Run this section only once.\n",
    "2. `Pipeline` - runs TRUST on a given BAM/fastaq file and analyzes the results. This section can be ran multiple times on different BAM/fastaq files.\n",
    "\n",
    "\n",
    "## Flow\n",
    "1. Run the sections `Functions` and `Configurations` once\n",
    "2. Run the section `'Pipeline` on a given BAM/fastaq file\n",
    "\n",
    "# Improvements\n",
    "- The functions (in the `Functions` section) should be implemented in a `./Utils/util.py` file, and imported here.\n",
    "- \n",
    "\n",
    "References:\n",
    "- [TRUST4 git hub](https://github.com/liulab-dfci/TRUST4)\n",
    "\n",
    "Notes:\n",
    "- The implementation below uses the TRUST4 docker image to run TRUST4. Thus, make sure docker is installed.\n",
    "\n",
    "TBD:\n",
    "- support for barcode path in `TRUSTContainer`?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import os\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions\n",
    "Should be moved to a `./Utils/utils.py` file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "type CommandOut = tuple[bool, str, str]\n",
    "\n",
    "def run_command(command: str) -> CommandOut:\n",
    "    \"\"\"Execute bash shell command.\n",
    "\n",
    "    Args:\n",
    "        command (str): shell command.\n",
    "\n",
    "    Returns:\n",
    "        tuple[bool, str, str]: exit_ok, stdout, stderr.\n",
    "    \"\"\"\n",
    "    ret = subprocess.run(command, shell=True, capture_output=True, check=False)\n",
    "    return ret.returncode==0, ret.stdout.decode(), ret.stderr.decode()\n",
    "\n",
    "class TRUSTContainer():\n",
    "    \"\"\"TRUST4 docker container.\"\"\"\n",
    "    def __init__(self, data_folder: str, ref_folder: str, output_folder: str, image: str, *, platform: str = 'linux/amd64') -> None:\n",
    "        self.image = image\n",
    "        self.bind_info: dict[str, str] = {'/ref': ref_folder, '/data': data_folder, '/out': output_folder}\n",
    "        self.base_command: str = f\"docker run --platform {platform} \" + ' '.join(f\"-v {v}:{k}\" for k, v in self.bind_info.items()) + f\" {self.image} run-trust4\"\n",
    "\n",
    "    def usage(self) -> CommandOut:\n",
    "        \"\"\"Shows usage\"\"\"\n",
    "        return run_command(self.base_command)\n",
    "\n",
    "    def _generate_command(self, ref_file: str, coordinate_file: str, extra_params: str | None = None) -> str:\n",
    "        \"\"\"command with --ref and -f options.\"\"\"\n",
    "        command = self.base_command + f\" --ref /ref/{ref_file} -f /ref/{coordinate_file} --od /out\"\n",
    "        if extra_params is not None:\n",
    "            command += f\" {extra_params}\"\n",
    "        return command\n",
    "\n",
    "    def generate_BAM_command(self, ref_file: str, coordinate_file: str, bam_file: str, extra_params: str | None = None) -> str:\n",
    "        \"\"\"generates the command to analyse a BAM file\"\"\"\n",
    "        return self._generate_command(ref_file, coordinate_file, extra_params=extra_params) + f\" -b /data/{bam_file}\"\n",
    "    \n",
    "    def generate_paired_fastaq_command(self, ref_file: str, coordinate_file: str, fasta_file1: str, fasta_file2: str, extra_params: str | None = None) -> str:\n",
    "        \"\"\"generates a command to analyze fastaq files/\"\"\"\n",
    "        return self._generate_command(ref_file, coordinate_file, extra_params=extra_params) + f\" -1 /data/{fasta_file1} -2 /data/{fasta_file2}\"\n",
    "    \n",
    "    def generate_single_fastaq_command(self, ref_file: str, coordinate_file: str, fasta_file: str, extra_params: str | None = None) -> str:\n",
    "        \"\"\"generates a command to analyze fastaq files/\"\"\"\n",
    "        return self._generate_command(ref_file, coordinate_file, extra_params=extra_params) + f\" -u /data/{fasta_file}\"\n",
    "\n",
    "    def run_BAM(self, ref_file: str, coordinate_file: str, bam_file: str, extra_params: str | None = None) -> CommandOut:\n",
    "        \"\"\"runs TRUST4 on BAM input file.\"\"\"\n",
    "        return run_command(\n",
    "            self.generate_BAM_command(ref_file, coordinate_file, bam_file, extra_params=extra_params)\n",
    "        )\n",
    "        #command = self.generate_BAM_command(ref_file, coordinate_file, bam_file, extra_params=extra_params)\n",
    "        #return run_command(command)\n",
    "\n",
    "    def run_paired_fastaq(self, ref_file: str, coordinate_file: str, fasta_file1: str, fasta_file2: str, extra_params: str | None = None) -> CommandOut:\n",
    "        \"\"\"runs TRUST4 on paired fastaq files.\"\"\"\n",
    "        return run_command(\n",
    "            self.generate_paired_fastaq_command(ref_file, coordinate_file, fasta_file1, fasta_file2, extra_params=extra_params)\n",
    "        )\n",
    "        #command = self.generate_paired_fastaq_command(ref_file, coordinate_file, fasta_file1, fasta_file2, extra_params=extra_params)\n",
    "        #return run_command(command)\n",
    "    \n",
    "    def run_single_fastaq(self, ref_file: str, coordinate_file: str, fasta_file: str, extra_params: str | None = None) -> CommandOut:\n",
    "        \"\"\"Runs TRUST4 on single fastaq file.\"\"\"\n",
    "        return run_command(\n",
    "            self.generate_single_fastaq_command(ref_file, coordinate_file, fasta_file, extra_params=extra_params)\n",
    "        )\n",
    "        #command = self.generate_single_fastaq_command(ref_file, coordinate_file, fasta_file, extra_params=extra_params)\n",
    "        #return run_command(command)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Configurations\n",
    "Run this section once to load and generate configurations files and the TRUST docker image."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameters\n",
    "# ==========\n",
    "\n",
    "# folder to hold reference files\n",
    "reference_path: Path = Path('./Reference')  # will hold all reference files\n",
    "\n",
    "# reference files\n",
    "genecode_file: Path = reference_path / 'genecode_annotation.gtf'  # annotation file\n",
    "hg38_ref_file: Path = reference_path / 'Homo_sapients_GRCh38.fa'  # hg38 reference genome fasta file\n",
    "VDJC_gene_ref_file: Path = reference_path / 'IMGT+C.fa'  # for the --ref file\n",
    "\n",
    "# BCR and TCR sequences and names\n",
    "bcr_tcr_gene_name_file: Path = reference_path / 'bcr_tcr_gene_name.txt'  # BCR and TCR gene names (required for generating the -f file)\n",
    "bcr_tcr_gene_seq_file: Path = reference_path / 'bcrtcr.fa'  # for the -f option (when using BAM files)\n",
    "\n",
    "# TRUST4 docker\n",
    "data_path: str = os.path.abspath('./Data')  # BAM files folders\n",
    "ref_path: str = os.path.abspath(reference_path)  # reference files folder\n",
    "out_path: str = os.path.abspath('./Output')  # TRUST output files folder\n",
    "docker_container: str = 'quay.io/biocontainers/trust4:1.1.5--h5ca1c30_0'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Configure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download genecode annotation file\n",
    "command = f\"wget -qO- https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_47/gencode.v47.annotation.gtf.gz | gunzip > {genecode_file}\"\n",
    "ret, stdout, stderr = run_command(command)\n",
    "print(ret)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download Ensembl GRCh38 genome\n",
    "command = f\"wget -qO- https://ftp.ensembl.org/pub/release-113/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz | gunzip > {hg38_ref_file}\"\n",
    "ret, stdout, stderr = run_command(command)\n",
    "print(ret)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the --ref file\n",
    "command = f\"perl BuildImgtAnnot.pl Homo_sapien > {VDJC_gene_ref_file}\"\n",
    "ret, stdout, stderr = run_command(command)\n",
    "print(ret)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate BCR and TCR gene names (needed for generating the -f file)\n",
    "command = f\"\"\"grep \">\" {VDJC_gene_ref_file} | cut -f2 -d'>' | cut -f1 -d'*' | sort | uniq > {bcr_tcr_gene_name_file}\"\"\"\n",
    "ret, stdout, stderr = run_command(command)\n",
    "print(ret)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# creare the BCR and TCR sequences (for the -f option)\n",
    "command = f\"perl BuildDatabaseFa.pl {hg38_ref_file} {genecode_file} {bcr_tcr_gene_name_file} > {bcr_tcr_gene_seq_file}\"\n",
    "ret, stdout, stderr = run_command(command)\n",
    "print(ret)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the TRUST4 docker image. Make sure docker is installed.\n",
    "command = f\"docker pull {docker_container}\"\n",
    "ret, stdout, stderr = run_command(command)\n",
    "print(ret)\n",
    "\n",
    "# verify\n",
    "!docker image ls | grep {docker_container.split(':')[0]}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Verify (Optional)\n",
    "We run `TRUST4` on the provided example BAM file to make sure TRUST4 is running properly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "\n",
      "TRUST4 v1.1.5-r573 usage: ./run-trust4 [OPTIONS]:\n",
      "Required:\n",
      "\t-b STRING: path to bam file\n",
      "\t-1 STRING -2 STRING: path to paired-end read files\n",
      "\t-u STRING: path to single-end read file\n",
      "\t-f STRING: path to the fasta file coordinate and sequence of V/D/J/C genes\n",
      "Optional:\n",
      "\t--ref STRING: path to detailed V/D/J/C gene reference file from IMGT database. (default: not used but recommended)\n",
      "\t-o STRING: prefix of output files. (default: inferred from file prefix)\n",
      "\t--od STRING: the directory for output files. (default: ./)\n",
      "\t-t INT: number of threads (default: 1)\n",
      "\t-k INT: the starting k-mer size for indexing contigs (default: 9)\n",
      "\t--barcode STRING: if -b, bam field for barcode; if -1 -2/-u, file containing barcodes (default: not used)\n",
      "\t--barcodeLevel STRING: barcode is for cell or molecule (default: cell)\n",
      "\t--barcodeWhitelist STRING: path to the barcode whitelist (default: not used)\n",
      "\t--barcodeTranslate STRING: path to the barcode translate file (default: not used)\n",
      "\t--UMI STRING: if -b, bam field for 10x Genomics-like UMI; if -1 -2/-u, file containing 10x Genomics-like UMIs (default: not used)\n",
      "\t--readFormat STRING: format for read, barcode and UMI files (example: r1:0:-1,r2:0:-1,bc:0:15,um:16:-1 for paired-end files with barcode and UMI)\n",
      "\t--repseq: the data is from bulk,non-UMI-based TCR-seq or BCR-seq (default: not set)\n",
      "\t--contigMinCov INT: ignore contigs that have bases covered by fewer than INT reads (default: 0)\n",
      "\t--minHitLen INT: the minimal hit length for a valid overlap (default: auto)\n",
      "\t--mateIdSuffixLen INT: the suffix length in read id for mate. (default: not used)\n",
      "\t--skipMateExtension: do not extend assemblies with mate information, useful for SMART-seq (default: not used)\n",
      "\t--skipReadRealign: do not realign reads in annotator, useful for reducing computation cost of barcode/UMI-based repseq (default: not used)\n",
      "\t--abnormalUnmapFlag: the flag in BAM for the unmapped read-pair is nonconcordant (default: not set)\n",
      "\t--noExtraction: directly use the files from provided -1 -2/-u to assemble (default: extraction first)\n",
      "\t--imgtAdditionalGap STRING: description for additional gap codon position in IMGT (0-based), e.g. \"TRAV:7,83\" for mouse (default: no)\n",
      "\t--assembleWithRef: conduct the assembly with --ref file (default: use -f file)\n",
      "\t--outputReadAssignment: output read assignment results to the prefix_assign.out file (default: no output)\n",
      "\t--stage INT: start TRUST4 on specified stage (default: 0):\n",
      "\t\t0: start from beginning (candidate read extraction)\n",
      "\t\t1: start from assembly\n",
      "\t\t2: start from annotation\n",
      "\t\t3: start from generating the report table\n",
      "\t--clean INT: clean up files. 0: no clean. 1: clean intermediate files. 2: only keep AIRR files. (default: 0)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# first show usage\n",
    "example_data_path: str = os.path.abspath('./example')\n",
    "example_ref_path: str = os.path.abspath('./')\n",
    "example_out_path: str = os.path.abspath('./example/Output')\n",
    "# -----------------------------------\n",
    "trust4c_exp = TRUSTContainer(example_data_path, example_ref_path, example_out_path, docker_container)\n",
    "ret, stdout, stderr = trust4c_exp.usage()\n",
    "print(ret, stdout, stderr, sep='\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "\n",
      "[Tue Apr  1 07:35:39 2025] TRUST4 v1.1.5-r573 begins.\n",
      "[Tue Apr  1 07:35:39 2025] SYSTEM CALL: /usr/local/bin/bam-extractor -b /data/example.bam -t 1 -f /ref/hg38_bcrtcr.fa -o /out/TRUST_example_toassemble \n",
      "[Tue Apr  1 07:35:39 2025] Start to extract candidate reads from bam file.\n",
      "[Tue Apr  1 07:35:39 2025] Finish obtaining the candidate read ids.\n",
      "[Tue Apr  1 07:35:39 2025] Finish extracting reads.\n",
      "[Tue Apr  1 07:35:39 2025] SYSTEM CALL: /usr/local/bin/trust4  -f /ref/hg38_bcrtcr.fa -o /out/TRUST_example -1 /out/TRUST_example_toassemble_1.fq -2 /out/TRUST_example_toassemble_2.fq\n",
      "[Tue Apr  1 07:35:40 2025] Start to assemble reads.\n",
      "[Tue Apr  1 07:35:40 2025] Found 394 reads.\n",
      "[Tue Apr  1 07:35:40 2025] Finish sorting the reads.\n",
      "[Tue Apr  1 07:35:40 2025] Finish rough annotations.\n",
      "[Tue Apr  1 07:35:40 2025] Assembled 394 reads.\n",
      "[Tue Apr  1 07:35:40 2025] Try to rescue 0 reads for assembly.\n",
      "[Tue Apr  1 07:35:40 2025] Rescued 0 reads.\n",
      "[Tue Apr  1 07:35:40 2025] Extend assemblies by mate pair information.\n",
      "[Tue Apr  1 07:35:40 2025] Remove redundant assemblies.\n",
      "[Tue Apr  1 07:35:41 2025] Finish assembly.\n",
      "[Tue Apr  1 07:35:41 2025] SYSTEM CALL: /usr/local/bin/annotator -f /ref/human_IMGT+C.fa -a /out/TRUST_example_final.out -t 1 -o /out/TRUST_example  -r /out/TRUST_example_assembled_reads.fa --airrAlignment > /out/TRUST_example_annot.fa\n",
      "[Tue Apr  1 07:35:41 2025] Start to annotate assemblies.\n",
      "[Tue Apr  1 07:35:42 2025] Start to realign reads for CDR3 analysis.\n",
      "[Tue Apr  1 07:35:42 2025] Compute CDR3 abundance.\n",
      "[Tue Apr  1 07:35:42 2025] Finish annotation.\n",
      "[Tue Apr  1 07:35:42 2025] SYSTEM CALL: perl /usr/local/bin/trust-simplerep.pl /out/TRUST_example_cdr3.out  > /out/TRUST_example_report.tsv\n",
      "[Tue Apr  1 07:35:42 2025] SYSTEM CALL: perl /usr/local/bin/trust-airr.pl /out/TRUST_example_report.tsv /out/TRUST_example_annot.fa --airr-align /out/TRUST_example_airr_align.tsv > /out/TRUST_example_airr.tsv\n",
      "[Tue Apr  1 07:35:42 2025] TRUST4 finishes.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# now run on example BAM file (with provided reference files)\n",
    "bam_file: Path = Path('example.bam')\n",
    "# -----------------------------------\n",
    "ret, stdout, stderr = trust4c_exp.run_BAM('human_IMGT+C.fa', 'hg38_bcrtcr.fa', bam_file.name)\n",
    "print(ret, stdout, stderr, sep='\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TRUST4 is ready to use.\n"
     ]
    }
   ],
   "source": [
    "# verify\n",
    "!bash mytrust_example_test.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now use the paired fastaq files\n",
    "ret, stdout, stderr = trust4c_exp.run_paired_fastaq('human_IMGT+C.fa', 'hg38_bcrtcr.fa', 'example_1.fq', 'example_2.fq')\n",
    "print(ret, stdout, stderr, sep='\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# verify\n",
    "!bash mytrust_example_test.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# show generated report\n",
    "report_file: Path = Path(example_out_path) / f'TRUST_{bam_file.stem}_report.tsv'\n",
    "df = pd.read_csv(report_file, sep='\\t').rename(columns={'#count': 'count'})\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now run with generated reference files\n",
    "trust4c = TRUSTContainer(data_path, ref_path, out_path, docker_container)\n",
    "ret, stdout, stderr = trust4c.run_BAM(VDJC_gene_ref_file.name, bcr_tcr_gene_seq_file.name, bam_file.name)\n",
    "print(ret, stdout, stderr, sep='\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# show example generated report\n",
    "report_file: Path = Path(out_path) / f'TRUST_{bam_file.stem}_report.tsv'\n",
    "df = pd.read_csv(report_file, sep='\\t').rename(columns={'#count': 'count'})\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis Pipeline\n",
    "\n",
    "### Data\n",
    "In the example below, we use several samples from this GEO [Dataset](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE262648): The tumor immune microenvironment and therapeutic efficacy of trastuzumab deruxtecan in gastric cancer.\n",
    "\n",
    "### Pre-requisite\n",
    "We are using the SRA toolkit to download the GEO samples:\n",
    "- download the [SRA toolkit](https://github.com/ncbi/sra-tools/wiki/01.-Downloading-SRA-Toolkit)\n",
    "- once download, execute `tar -xzvf sratoolkit.3.2.1-mac-arm64.tar`\n",
    "\n",
    "The SRA toolkit resides in `./Tools/sratoolkit.3.2.1/bin` folder.\n",
    "\n",
    "**Samples choosen**:\n",
    "- SRX24082037: GSM8172959: Sbj3_BL; Homo sapiens; RNA-Seq. Run accession number is SRR28479475\n",
    "- SRX24082038: GSM8172960: Sbj4_BL; Homo sapiens; RNA-Seq. Run accession number is SRR28479474\n",
    "\n",
    "**Downloading a sample**:\n",
    "1. Go to the GEO [Dataset](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE262648)\n",
    "1. Click on a sample from the Samples row (e.g., GSM8172959)\n",
    "1. Click on the SRA accession number at the bottom (under `Relations`), This will open the corresponding SRA page. Copy the corresponding SRA run accession (e.g. `SRR28479474`).\n",
    "1. To download the sample (e.g., `SRR28479474`), use the SRA toolkit as follows:\n",
    "\n",
    "    1. `prefetch SRR28479474`(or `prefetch SRX24082038`) - this downloads the sample in a `.sra` format\n",
    "    1. if the sample is not alligned (as in this case), perform: `fasterq-dump.3.2.1 ./SRR28479474 -t ./temp -O ./SRR28479474` - this will generate the fastaq file(s)\n",
    "    1. if the sample is aligned, perform `sam-dump.3.2.1 ./SRR28479474/SRR28479474.sra > ./SRR28479474/SRR28479474.sam` and then generate the BAM file by executing `samtools view -bS ./SRR28479474/SRR28479474.sam > ./SRR28479474/SRR28479474.bam`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# instantiate a TRUST4 class\n",
    "trust4c = TRUSTContainer(data_path, ref_path, out_path, docker_container)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "\n",
      "[Tue Apr  1 08:13:50 2025] TRUST4 v1.1.5-r573 begins.\n",
      "[Tue Apr  1 08:13:50 2025] SYSTEM CALL: /usr/local/bin/fastq-extractor -t 8 -f /ref/bcrtcr.fa -o /out/TRUST_SRR28479474_toassemble  -u /data/SRR28479474.fastq\n",
      "[Tue Apr  1 08:13:50 2025] Start to extract candidate reads from read files.\n",
      "[Tue Apr  1 08:26:25 2025] Finish extracting reads.\n",
      "[Tue Apr  1 08:26:26 2025] SYSTEM CALL: /usr/local/bin/trust4  -t 8 -f /ref/bcrtcr.fa -o /out/TRUST_SRR28479474 -u /out/TRUST_SRR28479474_toassemble.fq\n",
      "[Tue Apr  1 08:26:26 2025] Start to assemble reads.\n",
      "[Tue Apr  1 08:26:28 2025] Found 124528 reads.\n",
      "[Tue Apr  1 08:26:29 2025] Finish sorting the reads.\n",
      "[Tue Apr  1 08:26:33 2025] Finish rough annotations.\n",
      "[Tue Apr  1 08:26:38 2025] Processed 100000 reads (30295 are used for assembly).\n",
      "[Tue Apr  1 08:26:44 2025] Assembled 38399 reads.\n",
      "[Tue Apr  1 08:26:44 2025] Try to rescue 462 reads for assembly.\n",
      "[Tue Apr  1 08:26:44 2025] Rescued 89 reads.\n",
      "[Tue Apr  1 08:26:48 2025] SYSTEM CALL: /usr/local/bin/annotator -f /ref/IMGT+C.fa -a /out/TRUST_SRR28479474_final.out -t 8 -o /out/TRUST_SRR28479474  -r /out/TRUST_SRR28479474_assembled_reads.fa --airrAlignment > /out/TRUST_SRR28479474_annot.fa\n",
      "[Tue Apr  1 08:26:48 2025] Start to annotate assemblies.\n",
      "[Tue Apr  1 08:26:53 2025] Start to realign reads for CDR3 analysis.\n",
      "[Tue Apr  1 08:26:55 2025] Compute CDR3 abundance.\n",
      "[Tue Apr  1 08:26:56 2025] Finish annotation.\n",
      "[Tue Apr  1 08:26:56 2025] SYSTEM CALL: perl /usr/local/bin/trust-simplerep.pl /out/TRUST_SRR28479474_cdr3.out  > /out/TRUST_SRR28479474_report.tsv\n",
      "[Tue Apr  1 08:26:56 2025] SYSTEM CALL: perl /usr/local/bin/trust-airr.pl /out/TRUST_SRR28479474_report.tsv /out/TRUST_SRR28479474_annot.fa --airr-align /out/TRUST_SRR28479474_airr_align.tsv > /out/TRUST_SRR28479474_airr.tsv\n",
      "[Tue Apr  1 08:26:56 2025] TRUST4 finishes.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "''' \n",
    "Run the pipeline on a single-end fastaq file.\n",
    "\n",
    "In case of a BAM file, execute instead: \n",
    "    ret, stdout, stderr = trust4c.run_BAM(VDJC_gene_ref_file.name, bcr_tcr_gene_seq_file.name, bam_file, extra_params=f\" -t {num_threads}\")\n",
    "\n",
    "In case of paired fasta/fastq files, execute instead:\n",
    "    ret, stdout, stderr = trust4c.run_paired_fastaq(VDJC_gene_ref_file.name, bcr_tcr_gene_seq_file.name, fastaq_file1, fastaq_file2, extra_params=f\" -t {num_threads}\")\n",
    "\n",
    "'''\n",
    "#fastaq_file: str = 'SRR28479475.fastq'\n",
    "fastaq_file: str = 'SRR28479474.fastq'\n",
    "\n",
    "num_threads: int = 8\n",
    "# ------------------------------------\n",
    "# run_single_fastaq\n",
    "ret, stdout, stderr = trust4c.run_single_fastaq(VDJC_gene_ref_file.name, bcr_tcr_gene_seq_file.name, fastaq_file, extra_params=f\" -t {num_threads}\")\n",
    "print(ret, stdout, stderr, sep='\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Report from /Users/yoramzarai/work/Ramot/Projects/Exercises/TRUST4/Output/TRUST_SRR28479474_report.tsv:\n"
     ]
    },
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "count",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "frequency",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "CDR3nt",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "CDR3aa",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "V",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "D",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "J",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "C",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "cid",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "cid_full_length",
         "rawType": "int64",
         "type": "integer"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "72702005-1d4a-4de4-a067-cf9445f96c5d",
       "rows": [
        [
         "0",
         "58",
         "0.04035665",
         "TGCCAACAGTATAATAGTTATTCTATGTGGACGTTC",
         "CQQYNSYSMWTF",
         "IGKV1-5*03",
         ".",
         "IGKJ1*01",
         "IGKC",
         "assemble12",
         "0"
        ],
        [
         "1",
         "50",
         "0.03429355",
         "TGCGGAACATGGGAAAGCAGCCTGAGTGCTGGGGTGTTC",
         "CGTWESSLSAGVF",
         "IGLV1-51*01",
         ".",
         "IGLJ3*02",
         "IGLC",
         "assemble75",
         "0"
        ],
        [
         "2",
         "48",
         "0.1804511",
         "TGTGCTAGAGATGGCGCCCCACATTTTAACTGGTTACTTGACCACTGG",
         "CARDGAPHFNWLLDHW",
         "IGHV4-39*07",
         "IGHD3-9*01",
         "IGHJ4*02",
         ".",
         "assemble100",
         "0"
        ],
        [
         "3",
         "47",
         "0.03284636",
         "TGTCAACAGAGTTACGGTAGTCCTCCCACTTTT",
         "CQQSYGSPPTF",
         "IGKV1-39*01",
         ".",
         "IGKJ2*01",
         "IGKC",
         "assemble60",
         "0"
        ],
        [
         "4",
         "42",
         "0.02917695",
         "TGTCAACAGAGTTACAGTCCCCTTTACACTTTT",
         "CQQSYSPLYTF",
         "IGKV1-39*01",
         ".",
         "IGKJ2*01",
         "IGKC",
         "assemble38",
         "0"
        ],
        [
         "5",
         "38",
         "0.0260631",
         "TGCATGCAAGCTCTACAAACTCTGTACACTTTT",
         "CMQALQTLYTF",
         "IGKV2-28*01",
         ".",
         "IGKJ2*01",
         "IGKC",
         "assemble47",
         "0"
        ],
        [
         "6",
         "34",
         "0.1278195",
         "TGTGCGAGAGGTCCAAGGACTTCCTTGGAACTTGACTACTGG",
         "CARGPRTSLELDYW",
         "IGHV3-21*06",
         "IGHD1-7*01",
         "IGHJ4*02",
         ".",
         "assemble86",
         "0"
        ],
        [
         "7",
         "30",
         "0.02057613",
         "TGTCAGCAGCGTAGCAACTGGCCCCTCACTTTC",
         "CQQRSNWPLTF",
         "IGKV3D-11*03",
         ".",
         "IGKJ4*01",
         "IGKC",
         "assemble63",
         "0"
        ],
        [
         "8",
         "29",
         "0.02030864",
         "TGCCAGTCTTATGACATCAGCCTGAGTGGTGTGGTATTC",
         "CQSYDISLSGVVF",
         "IGLV1-40*01",
         ".",
         "IGLJ2*01",
         "IGLC",
         "assemble79",
         "0"
        ],
        [
         "9",
         "29",
         "0.01989026",
         "TGTCACCAGTATGGTAGTCCATTCACTTTC",
         "CHQYGSPFTF",
         "IGKV3-20*01",
         ".",
         "IGKJ3*01",
         "IGKC",
         "assemble187",
         "0"
        ],
        [
         "10",
         "28",
         "0.01940329",
         "TGCATGCAACGTTTAGACTTTCCGATCACCTTC",
         "CMQRLDFPITF",
         "IGKV2-40*01",
         ".",
         "IGKJ5*01",
         "IGKC",
         "assemble172",
         "0"
        ],
        [
         "11",
         "28",
         "0.01920439",
         "TGTCAGCAGGGTAGCAACTGGCCGCTCACTTTC",
         "CQQGSNWPLTF",
         "IGKV3-11*01",
         ".",
         "IGKJ4*01",
         "IGKC",
         "assemble285",
         "0"
        ],
        [
         "12",
         "28",
         "0.01920439",
         "TCTCAGCAGTATAATAACTGGCCAAGGACGTTC",
         "SQQYNNWPRTF",
         "IGKV3-15*01",
         ".",
         "IGKJ1*01",
         "IGKC",
         "assemble64",
         "0"
        ],
        [
         "13",
         "27",
         "0.1015038",
         "TGTGCGAGACATGGTCCACACCCGACCACGGTGATTCCTTCATTTGACTACCGG",
         "CARHGPHPTTVIPSFDYR",
         "IGHV4-39*01",
         "IGHD4-17*01",
         "IGHJ4*02",
         "IGHA1",
         "assemble4",
         "0"
        ],
        [
         "14",
         "25",
         "0.01770233",
         "TGCTGCTCATATGCAGGGGACTACACTTTAGTTGTCTTC",
         "CCSYAGDYTLVVF",
         "IGLV2-11*01",
         ".",
         "IGLJ1*01",
         "IGLC",
         "assemble93",
         "0"
        ],
        [
         "15",
         "24",
         "0.01646091",
         "TGCATGCAAACTCTAGAAACTCCGTTCACTTTT",
         "CMQTLETPFTF",
         "IGKV2-28*01",
         ".",
         "IGKJ2*01",
         "IGKC",
         "assemble81",
         "0"
        ],
        [
         "16",
         "24",
         "0.01646091",
         "TGTCAGCAGTATGATAACTGGCCTCTCACTTTC",
         "CQQYDNWPLTF",
         "IGKV3-15*01",
         ".",
         "IGKJ4*01",
         "IGKC",
         "assemble49",
         "0"
        ],
        [
         "17",
         "21",
         "0.01440329",
         "TGCCAACAGTATAGTAGTTATTCTCCGTCCACTTTT",
         "CQQYSSYSPSTF",
         "IGKV1-5*03",
         ".",
         "IGKJ2*01",
         "IGKC",
         "assemble503",
         "0"
        ],
        [
         "18",
         "21",
         "0.07894737",
         "TGTGCGGGCGGAACTATAACTCCAACTTACTACTTCCCTATGGACGTCTGG",
         "CAGGTITPTYYFPMDVW",
         "IGHV4-59*09",
         "IGHD1-7*01",
         "IGHJ6*02",
         "IGHG1",
         "assemble412",
         "0"
        ],
        [
         "19",
         "20",
         "0.7692308",
         "TGTTCCAGCAGACAGGGACAGGGGGCAGATACGCAGTATTTT",
         "CSSRQGQGADTQYF",
         "TRBV12-3*01",
         "TRBD1*01",
         "TRBJ2-3*01",
         "TRBC",
         "assemble118",
         "0"
        ],
        [
         "20",
         "20",
         "0.07518797",
         "TGTGCGAGACATCAATTTTATGGTTCGGGGAGTTATCCGGACTACTGG",
         "CARHQFYGSGSYPDYW",
         "IGHV4-39*01",
         "IGHD3-10*01",
         "IGHJ4*02",
         ".",
         "assemble743",
         "0"
        ],
        [
         "21",
         "19",
         "0.01349108",
         "TGTCTACAGCATGATAGTTACCCTCTGACGTTC",
         "CLQHDSYPLTF",
         "IGKV1-17*01",
         ".",
         "IGKJ1*01",
         "IGKC",
         "assemble298",
         "0"
        ],
        [
         "22",
         "18",
         "0.01290809",
         "TGTCAACAGTATGATAAGTCAGTCTCTTTT",
         "CQQYDKSVSF",
         "IGKV3-20*01",
         ".",
         "IGKJ2*01",
         "IGKC",
         "assemble122",
         "0"
        ],
        [
         "23",
         "18",
         "0.01234568",
         "TGTCAGCAGTATGGTAGCTCACCGCTCACTTTC",
         "CQQYGSSPLTF",
         "IGKV3-20*01",
         ".",
         "IGKJ4*01",
         "IGKC",
         "assemble105",
         "0"
        ],
        [
         "24",
         "17",
         "0.01165981",
         "TGCCAACAGTATAATGCTTACCCTTACACTTTT",
         "CQQYNAYPYTF",
         "IGKV1-16*01",
         ".",
         "IGKJ2*01",
         "IGKC",
         "assemble160",
         "0"
        ],
        [
         "25",
         "16",
         "0.06349624",
         "TGTGCGAAAGACTATACCAGCAGATACTATTACGCTATGCACGTCTGG",
         "CAKDYTSRYYYAMHVW",
         "IGHV1-69*01",
         "IGHD6-25*01",
         "IGHJ6*02",
         ".",
         "assemble150",
         "0"
        ],
        [
         "26",
         "16",
         "0.01097394",
         "TGTCAACACAGTTACAATATTCCGTGGACGTTC",
         "CQHSYNIPWTF",
         "IGKV1-39*01",
         ".",
         "IGKJ1*01",
         "IGKC",
         "assemble173",
         "0"
        ],
        [
         "27",
         "16",
         "0.01097394",
         "TGTCAGTACTATGTTAGCCCACTTACGTGGACGTTC",
         "CQYYVSPLTWTF",
         "IGKV3-20*01",
         ".",
         "IGKJ1*01",
         "IGKC",
         "assemble398",
         "0"
        ],
        [
         "28",
         "15",
         "0.01043896",
         "TGCCAAAGGTGTAATACTTACCCGTACACTTTT",
         "CQRCNTYPYTF",
         "IGKV1-13*01",
         ".",
         "IGKJ2*01",
         "IGKC",
         "assemble108",
         "0"
        ],
        [
         "29",
         "15",
         "0.05639098",
         "TGTGCGAGAATAGGGGGAGGTTTGGTGGGAGCTAATGAGACCCCACCCATCTGG",
         "CARIGGGLVGANETPPIW",
         "IGHV1-46*01",
         "IGHD3-16*01",
         "IGHJ3*02",
         "IGHG1",
         "assemble91",
         "0"
        ],
        [
         "30",
         "15",
         "0.01028807",
         "TGTCAGGTATGGGATAGTAATAGTGATCATCATGTCTTC",
         "CQVWDSNSDHHVF",
         "IGLV3-21*02",
         ".",
         "IGLJ1*01",
         "IGLC",
         "assemble697",
         "0"
        ],
        [
         "31",
         "14",
         "0.009718793",
         "TGTCAAAACTATGATAATCGCGGAGTGTTCACATTC",
         "CQNYDNRGVFTF",
         "IGKV1-33*01",
         ".",
         "IGKJ4*01",
         "IGKC",
         "assemble65",
         "0"
        ],
        [
         "32",
         "14",
         "0.009602195",
         "TGCATGCAAGCTCTACAAACTCCGTGGACGTTC",
         "CMQALQTPWTF",
         "IGKV2-28*01",
         ".",
         "IGKJ1*01",
         "IGKC",
         "assemble227",
         "0"
        ],
        [
         "33",
         "13",
         "0.008916324",
         "TGCATGCAAGCTACACAATTTCCTCGGACTTTT",
         "CMQATQFPRTF",
         "IGKV2-24*01",
         ".",
         "IGKJ2*01",
         "IGKC",
         "assemble202",
         "0"
        ],
        [
         "34",
         "12",
         "0.04511278",
         "TGTGTAAGAGACCCCGACTACCTAGACTATGTCGGCAACGGGATGGACGTCTGG",
         "CVRDPDYLDYVGNGMDVW",
         "IGHV1-18*01",
         "IGHD4-17*01",
         "IGHJ6*02",
         ".",
         "assemble1125",
         "0"
        ],
        [
         "35",
         "11",
         "0.007805213",
         "TGTCAACAGAGTTACAGTACCCCCGGGACGTTC",
         "CQQSYSTPGTF",
         "IGKV1-39*01",
         ".",
         "IGKJ1*01",
         "IGKC",
         "assemble1080",
         "0"
        ],
        [
         "36",
         "11",
         "0.44",
         "TGTGCAATGAGAGAGGGCCAGAACACCGACAAGCTCATCTTT",
         "CAMREGQNTDKLIF",
         "TRAV14/DV4*01",
         ".",
         "TRAJ34*01",
         ".",
         "assemble435",
         "0"
        ],
        [
         "37",
         "11",
         "0.007544582",
         "TGTCAACAATATAATATATATTCGACGTTC",
         "CQQYNIYSTF",
         "IGKV1-5*01",
         ".",
         "IGKJ1*01",
         "IGKC",
         "assemble490",
         "0"
        ],
        [
         "38",
         "11",
         "0.04135338",
         "TGTGTGAGGAACACTCTGGTTCGGGGAGTAGTGAGCTACTACCACGGCATGGACGTCTGG",
         "CVRNTLVRGVVSYYHGMDVW",
         "IGHV3-48*03",
         "IGHD3-10*01",
         "IGHJ6*02",
         ".",
         "assemble691",
         "0"
        ],
        [
         "39",
         "11",
         "0.007544582",
         "TGCAGCTGCTGCTCATATACAAGGACCACCACTCTCGTGGTATTC",
         "CSCCSYTRTTTLVVF",
         "IGLV2-14*01",
         ".",
         "IGLJ2*01",
         "IGLC",
         "assemble269",
         "0"
        ],
        [
         "40",
         "11",
         "0.04135338",
         "TGTGCGAGAGCAACGTGGTGGTTATTGTCTTGACTACTGG",
         "out_of_frame",
         "IGHV1-46*01",
         "IGHD2-21*01",
         "IGHJ4*01",
         ".",
         "assemble663",
         "0"
        ],
        [
         "41",
         "10",
         "0.006858711",
         "TGTCAACACCATAATAGTTTCCCTCCCTTACTCACTTTC",
         "CQHHNSFPPLLTF",
         "IGKV1D-8*03",
         ".",
         "IGKJ3*01",
         "IGKC",
         "assemble1797",
         "0"
        ],
        [
         "42",
         "10",
         "0.006858711",
         "TGTATGCAAACTCTAGAAACTCCGTTCACTTTT",
         "CMQTLETPFTF",
         "IGKV2-28*01",
         ".",
         "IGKJ2*01",
         "IGKC",
         "assemble81",
         "0"
        ],
        [
         "43",
         "10",
         "0.006858711",
         "TGTCAACAGAGTTACAGTACCCTCCTCATTTTC",
         "CQQSYSTLLIF",
         "IGKV1-39*01",
         ".",
         "IGKJ3*01",
         "IGKC",
         "assemble428",
         "0"
        ],
        [
         "44",
         "10",
         "0.006858711",
         "TGTATGATTTGGCACAGCAGCGCCTGGGTCTTC",
         "CMIWHSSAWVF",
         "IGLV5-45*01",
         ".",
         "IGLJ3*02",
         "IGLC",
         "assemble349",
         "0"
        ],
        [
         "45",
         "9",
         "0.36",
         "TGTGCAGCCCTTAGGCCTAACGACTACAAGCTCAGCTTT",
         "CAALRPNDYKLSF",
         "TRAV13-1*01",
         ".",
         "TRAJ20*01",
         ".",
         "assemble377",
         "0"
        ],
        [
         "46",
         "9",
         "0.03383459",
         "TGTGCGAGAGATTGTCGGGGGGAAACAACCAGTGAAAATACCGTGGTGATGGACGTCTGG",
         "CARDCRGETTSENTVVMDVW",
         "IGHV1-18*01",
         "IGHD2-21*01",
         "IGHJ6*02",
         ".",
         "assemble1392",
         "0"
        ],
        [
         "47",
         "9",
         "0.00617284",
         "TGTCAGCAGCGTAGCAACTGGCTCGCGCTCAGTTTC",
         "CQQRSNWLALSF",
         "IGKV3-11*01",
         ".",
         "IGKJ4*01",
         "IGKC",
         "assemble1451",
         "0"
        ],
        [
         "48",
         "9",
         "0.00617284",
         "TGTCAGCAGCATAATAACTGGCCAACTTTC",
         "CQQHNNWPTF",
         "IGKV3-15*01",
         ".",
         "IGKJ3*01",
         "IGKC",
         "assemble1097",
         "0"
        ],
        [
         "49",
         "9",
         "0.00617284",
         "TGTCAGCAGTATGGTAGCTCACCGCTCTTTC",
         "out_of_frame",
         "IGKV3-20*01",
         ".",
         "IGKJ4*01",
         "IGKC",
         "assemble371",
         "0"
        ]
       ],
       "shape": {
        "columns": 10,
        "rows": 340
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>frequency</th>\n",
       "      <th>CDR3nt</th>\n",
       "      <th>CDR3aa</th>\n",
       "      <th>V</th>\n",
       "      <th>D</th>\n",
       "      <th>J</th>\n",
       "      <th>C</th>\n",
       "      <th>cid</th>\n",
       "      <th>cid_full_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>58</td>\n",
       "      <td>0.040357</td>\n",
       "      <td>TGCCAACAGTATAATAGTTATTCTATGTGGACGTTC</td>\n",
       "      <td>CQQYNSYSMWTF</td>\n",
       "      <td>IGKV1-5*03</td>\n",
       "      <td>.</td>\n",
       "      <td>IGKJ1*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>assemble12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>50</td>\n",
       "      <td>0.034294</td>\n",
       "      <td>TGCGGAACATGGGAAAGCAGCCTGAGTGCTGGGGTGTTC</td>\n",
       "      <td>CGTWESSLSAGVF</td>\n",
       "      <td>IGLV1-51*01</td>\n",
       "      <td>.</td>\n",
       "      <td>IGLJ3*02</td>\n",
       "      <td>IGLC</td>\n",
       "      <td>assemble75</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>48</td>\n",
       "      <td>0.180451</td>\n",
       "      <td>TGTGCTAGAGATGGCGCCCCACATTTTAACTGGTTACTTGACCACTGG</td>\n",
       "      <td>CARDGAPHFNWLLDHW</td>\n",
       "      <td>IGHV4-39*07</td>\n",
       "      <td>IGHD3-9*01</td>\n",
       "      <td>IGHJ4*02</td>\n",
       "      <td>.</td>\n",
       "      <td>assemble100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>47</td>\n",
       "      <td>0.032846</td>\n",
       "      <td>TGTCAACAGAGTTACGGTAGTCCTCCCACTTTT</td>\n",
       "      <td>CQQSYGSPPTF</td>\n",
       "      <td>IGKV1-39*01</td>\n",
       "      <td>.</td>\n",
       "      <td>IGKJ2*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>assemble60</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>42</td>\n",
       "      <td>0.029177</td>\n",
       "      <td>TGTCAACAGAGTTACAGTCCCCTTTACACTTTT</td>\n",
       "      <td>CQQSYSPLYTF</td>\n",
       "      <td>IGKV1-39*01</td>\n",
       "      <td>.</td>\n",
       "      <td>IGKJ2*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>assemble38</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000686</td>\n",
       "      <td>TTTCAACAGAGTTACAGTACCCCCGGGACGTTC</td>\n",
       "      <td>FQQSYSTPGTF</td>\n",
       "      <td>IGKV1-39*01</td>\n",
       "      <td>.</td>\n",
       "      <td>IGKJ1*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>assemble4578</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000686</td>\n",
       "      <td>TGTGTGGTGTATATGGGTGGTGGCATTTGGGTGTTC</td>\n",
       "      <td>CVVYMGGGIWVF</td>\n",
       "      <td>IGLV8-61*01</td>\n",
       "      <td>.</td>\n",
       "      <td>IGLJ3*02</td>\n",
       "      <td>IGLC</td>\n",
       "      <td>assemble1656</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000686</td>\n",
       "      <td>TGTTCAGCATGGGATGACTACCTGCATGGTCTGTTC</td>\n",
       "      <td>CSAWDDYLHGLF</td>\n",
       "      <td>IGLV1-44*01</td>\n",
       "      <td>.</td>\n",
       "      <td>IGLJ3*02</td>\n",
       "      <td>IGLC</td>\n",
       "      <td>assemble4710</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>1</td>\n",
       "      <td>0.003759</td>\n",
       "      <td>TGTGCACGGATACGCGGGGATATTTTGACTGGTTATTATAACTTTG...</td>\n",
       "      <td>CARIRGDILTGYYNFDAFDIW</td>\n",
       "      <td>IGHV2-70*01</td>\n",
       "      <td>IGHD3-9*01</td>\n",
       "      <td>IGHJ3*02</td>\n",
       "      <td>.</td>\n",
       "      <td>assemble4839</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000686</td>\n",
       "      <td>TGTCATCGTGACAACTGGCCTCTCACTTTC</td>\n",
       "      <td>CHRDNWPLTF</td>\n",
       "      <td>IGKV3-11*01</td>\n",
       "      <td>.</td>\n",
       "      <td>IGKJ4*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>assemble3959</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>340 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     count  frequency                                             CDR3nt  \\\n",
       "0       58   0.040357               TGCCAACAGTATAATAGTTATTCTATGTGGACGTTC   \n",
       "1       50   0.034294            TGCGGAACATGGGAAAGCAGCCTGAGTGCTGGGGTGTTC   \n",
       "2       48   0.180451   TGTGCTAGAGATGGCGCCCCACATTTTAACTGGTTACTTGACCACTGG   \n",
       "3       47   0.032846                  TGTCAACAGAGTTACGGTAGTCCTCCCACTTTT   \n",
       "4       42   0.029177                  TGTCAACAGAGTTACAGTCCCCTTTACACTTTT   \n",
       "..     ...        ...                                                ...   \n",
       "335      1   0.000686                  TTTCAACAGAGTTACAGTACCCCCGGGACGTTC   \n",
       "336      1   0.000686               TGTGTGGTGTATATGGGTGGTGGCATTTGGGTGTTC   \n",
       "337      1   0.000686               TGTTCAGCATGGGATGACTACCTGCATGGTCTGTTC   \n",
       "338      1   0.003759  TGTGCACGGATACGCGGGGATATTTTGACTGGTTATTATAACTTTG...   \n",
       "339      1   0.000686                     TGTCATCGTGACAACTGGCCTCTCACTTTC   \n",
       "\n",
       "                    CDR3aa            V           D         J     C  \\\n",
       "0             CQQYNSYSMWTF   IGKV1-5*03           .  IGKJ1*01  IGKC   \n",
       "1            CGTWESSLSAGVF  IGLV1-51*01           .  IGLJ3*02  IGLC   \n",
       "2         CARDGAPHFNWLLDHW  IGHV4-39*07  IGHD3-9*01  IGHJ4*02     .   \n",
       "3              CQQSYGSPPTF  IGKV1-39*01           .  IGKJ2*01  IGKC   \n",
       "4              CQQSYSPLYTF  IGKV1-39*01           .  IGKJ2*01  IGKC   \n",
       "..                     ...          ...         ...       ...   ...   \n",
       "335            FQQSYSTPGTF  IGKV1-39*01           .  IGKJ1*01  IGKC   \n",
       "336           CVVYMGGGIWVF  IGLV8-61*01           .  IGLJ3*02  IGLC   \n",
       "337           CSAWDDYLHGLF  IGLV1-44*01           .  IGLJ3*02  IGLC   \n",
       "338  CARIRGDILTGYYNFDAFDIW  IGHV2-70*01  IGHD3-9*01  IGHJ3*02     .   \n",
       "339             CHRDNWPLTF  IGKV3-11*01           .  IGKJ4*01  IGKC   \n",
       "\n",
       "              cid  cid_full_length  \n",
       "0      assemble12                0  \n",
       "1      assemble75                0  \n",
       "2     assemble100                0  \n",
       "3      assemble60                0  \n",
       "4      assemble38                0  \n",
       "..            ...              ...  \n",
       "335  assemble4578                0  \n",
       "336  assemble1656                0  \n",
       "337  assemble4710                0  \n",
       "338  assemble4839                0  \n",
       "339  assemble3959                0  \n",
       "\n",
       "[340 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# show report\n",
    "report_file: Path = Path(out_path) / f'TRUST_{Path(fastaq_file).stem}_report.tsv'\n",
    "df = pd.read_csv(report_file, sep='\\t').rename(columns={'#count': 'count'})\n",
    "print(f\"Report from {report_file}:\")\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ScratchPad"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Things we tried:\n",
    "\n",
    "1. download the SRA toolkit from [here](https://github.com/ncbi/sra-tools/wiki/01.-Downloading-SRA-Toolkit)\n",
    "1. once download, execute `tar -xzvf sratoolkit.3.2.1-mac-arm64.tar`\n",
    "1. dataset from Immunophenotyping of COVID-19 and Influenza Underscores the Association of Type I IFN Response in Severe COVID-19 in [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE149689): \n",
    "\n",
    "From the downloaded folder execute:\n",
    "1. ./bin/prefetch SRX8241106 --max-size 30G\n",
    "\n",
    "Resulted with:\n",
    "```console\n",
    "2025-03-31T13:15:00 prefetch.3.2.1: 1) Resolving 'SRX8241106'...\n",
    "2025-03-31T13:15:03 prefetch.3.2.1: Current preference is set to retrieve SRA Normalized Format files with full base quality scores\n",
    "2025-03-31T13:15:03 prefetch.3.2.1: 1) Downloading 'SRR11680207'...\n",
    "2025-03-31T13:15:03 prefetch.3.2.1:  SRA Normalized Format file is being retrieved\n",
    "2025-03-31T13:15:03 prefetch.3.2.1:  Downloading via HTTPS...\n",
    "2025-03-31T13:52:10 prefetch.3.2.1:  HTTPS download succeed\n",
    "2025-03-31T13:52:10 prefetch.3.2.1:   verifying 'SRR11680207'...\n",
    "2025-03-31T13:52:57 prefetch.3.2.1:  'SRR11680207' is valid: 29607783613 bytes were streamed from 29607781179\n",
    "2025-03-31T13:52:57 prefetch.3.2.1: 1) 'SRR11680207' was downloaded successfully\n",
    "2025-03-31T13:52:57 prefetch.3.2.1: 1) Resolving 'SRX8241106's dependencies...\n",
    "2025-03-31T13:52:57 prefetch.3.2.1: 'SRX8241106' has 0 unresolved dependencies\n",
    "```\n",
    "\n",
    "1. then generate the fastaq files: ./bin/fasterq-dump.3.2.1 ./SRR11680207 -t ./temp -O ./SRR11680207\n",
    "\n",
    "resulted in:\n",
    "```console\n",
    "spots read      : 722,551,560\n",
    "reads read      : 1,445,103,120\n",
    "reads written   : 1,445,103,120\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "Running on the two fastaq files:\n",
    "```python\n",
    "fastq_1_file = 'SRR11680207_1.fastq'\n",
    "fastq_2_file = 'SRR11680207_2.fastq'\n",
    "\n",
    "num_threads: int = 8\n",
    "# -----------------------\n",
    "ret, stdout, stderr = trust4c.run_paired_fastaq(VDJC_gene_ref_file.name, bcr_tcr_gene_seq_file.name, fastq_1_file, fastq_2_file, extra_params=f\" -t {num_threads}\")\n",
    "print(ret, stdout, stderr, sep='\\n')\n",
    "```\n",
    "\n",
    "Resulted in:\n",
    "```console\n",
    "Mon Mar 31 16:43:23 2025] TRUST4 v1.1.5-r573 begins.\n",
    "[Mon Mar 31 16:43:23 2025] SYSTEM CALL: /usr/local/bin/fastq-extractor -t 8 -f /ref/bcrtcr.fa -o /out/TRUST_SRR11680207_1_toassemble  -1 /data/SRR11680207_1.fastq -2 /data/SRR11680207_2.fastq\n",
    "[Mon Mar 31 16:43:23 2025] Start to extract candidate reads from read files.\n",
    "system /usr/local/bin/fastq-extractor -t 8 -f /ref/bcrtcr.fa -o /out/TRUST_SRR11680207_1_toassemble  -1 /data/SRR11680207_1.fastq -2 /data/SRR11680207_2.fastq failed: 9 at /usr/local/bin/run-trust4 line 60\n",
    "```\n",
    "\n",
    "Tried also to create the BAM file:\n",
    "1. ./bin/sam-dump.3.2.1 ./SRR11680207/SRR11680207.sra > ./SRR11680207/SRR11680207.sam\n",
    "1. samtools view -bS ./SRR11680207/SRR11680207.sam > ./SRR11680207/SRR11680207.bam\n",
    "\n",
    "And then running on the BAM file resulted in a similar error.\n",
    "\n",
    "Try now https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE262648:\n",
    "1. ./bin/prefetch SRR28479475\n",
    "1. ./bin/sam-dump.3.2.1 ./SRR28479475/SRR28479475.sra > ./SRR28479475/SRR28479475.sam\n",
    "1. samtools view -bS ./SRR28479475/SRR28479475.sam > ./SRR28479475/SRR28479475.bam"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
