{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TRUST4 Exercise\n",
    "\n",
    "This notebook contains the implementation code and descriptions of the Numenos's TRUST4 excercise. \n",
    "\n",
    "The notebook contains the following main sections:\n",
    "1. `Parameters` - sets the different parameters\n",
    "1. `Configurations` - installs and generates reference data, and installs the TRUST4 docker image\n",
    "1. `Pipeline` - runs TRUST4 on a given BAM/fastaq file. The section can be executed multiple times on different BAM/fastaq files\n",
    "1. `Result Evaluations` - analyzes the TRUST4 results\n",
    "1. `Results Summary` - summarizes main observations\n",
    "1. `Download GEO Samples` - downloads samples using the SRA tookit\n",
    "\n",
    "### TRUST4\n",
    "- an efficient tool to infer TCR and BCR repertoires from bulk RNA-seq or scRNA-seq data\n",
    "- identifies potential TCR and BCR reads from raw sequence files (detects significant overlaps with V, D, J, and C genes)\n",
    "- assembles candidate reads into contigs (prioritizes reads from highly expressed receptor genes)\n",
    "- annotates contigs (using IMGT) and identifies V, D, J, and C genes\n",
    "- determines complementary-determining region 3 (CDR3) coordinates based on IMGT\n",
    "\n",
    "### Execution Flow\n",
    "1. Run the sections `Functions and Classes`, `Parameters`, and `Configurations` once\n",
    "2. Run the section `Pipeline` on a given BAM/fastaq file followed by the section `Result Evaluations`\n",
    "\n",
    "### Challenges Encountered \n",
    "- Downloading samples from [GEO](https://www.google.com/search?client=safari&rls=en&q=GEO&ie=UTF-8&oe=UTF-8) requires (in most cases) an AWS or a Google bucket, or using the [SRA toolkit](https://github.com/ncbi/sra-tools/wiki/01.-Downloading-SRA-Toolkit). If using the SRA toolkit, one needs to know if the sample is aligned (to generate a BAM file) or un-aligned (to generate the fastaq files(s)). This information may be present in the sample page. Otherwise, use the `sra-stat --xml <sample> > info.xml` command from the SRA toolkit and look in the `<AlignInfo>` field, or the `vdb-dump --info <sample>` command\n",
    "\n",
    "### QC Considerations\n",
    "- first, make sure TRUST4 runs successfully on the provided example (compare to expected output)\n",
    "- expected higher number of BCRs than TCRs \n",
    "- consistency among output files (i.e., `report.tsv`, `cdr3.out`, `annot.fa`) based on consensus ID\n",
    "- consistency in number of detected genes and CDR3 among samples\n",
    "- low percentage of contigs that contain the complete variable domain of the receptor (`cid_full_length` == 1) (?)\n",
    "- low average read counts observed (?)\n",
    "- filter out V(D)J recombination with low read counts\n",
    "- comparing the diversity of the V, J, D, and C genes reported with known gene usage patterns\n",
    "- compare TRUST4's results with other established reconstruction tools\n",
    "\n",
    "### Improvements\n",
    "- the pipeline is implemented in a Jupyter notebook in order to ease the review process (since it is an evaluation exercise). In general, the pipeline should be implemented in python scipts to accommodate efficient workflows and executions (e.g., job scheduling)\n",
    "- the functions and classes should be implemented in `./Utils/command.py`, `./Utils/TRUSTdocker_api.py`, and `./Utils/SRAtoolkit_api.py` files\n",
    "- configuration and pipeline code should be separated\n",
    "- a configuration file (toml or json) should be used\n",
    "- tests should be added\n",
    "- workflow management framework should be considered\n",
    "\n",
    "### Notes\n",
    "- The implementation below downloads and uses the `TRUST4 docker` image to run TRUST4. Thus, make sure `docker` is installed\n",
    "- Make sure perl5 is installed\n",
    "\n",
    "### References:\n",
    "- [TRUST4 git hub](https://github.com/liulab-dfci/TRUST4)\n",
    "- [SRA toolkit](https://github.com/ncbi/sra-tools/wiki/01.-Downloading-SRA-Toolkit)\n",
    "- [GEO](https://www.ncbi.nlm.nih.gov/geo/)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    "1. [Functions and Classes](#FunctionsandClasses)\n",
    "1. [Parameters](#Parameters)\n",
    "1. [Configurations](#Configurations)\n",
    "1. [Pipeline](#Pipeline)\n",
    "1. [Result Evaluations](#resulteval)\n",
    "1. [Results Summary](#resultsum)\n",
    "1. [Download GEO Samples](#donaloaddata)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "from collections import defaultdict\n",
    "from pathlib import Path\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"FunctionsandClasses\"></a>\n",
    "## Functions and Classes\n",
    "TRUST4 docker image python interface."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "type CommandOut = tuple[bool, str, str]  # command output type\n",
    "\n",
    "def run_command(command: str, check: bool = True) -> CommandOut:\n",
    "    \"\"\"Execute bash shell command.\n",
    "\n",
    "    Args:\n",
    "        command (str): shell command.\n",
    "\n",
    "    Returns:\n",
    "        tuple[bool, str, str]: exit_ok, stdout, stderr.\n",
    "    \"\"\"\n",
    "    ret = subprocess.run(command, shell=True, capture_output=True, check=check)\n",
    "    return ret.returncode==0, ret.stdout.decode(), ret.stderr.decode()\n",
    "\n",
    "\n",
    "class TRUSTContainer():\n",
    "    \"\"\"TRUST4 docker container interface.\"\"\"\n",
    "    def __init__(self, data_folder: str, ref_folder: str, output_folder: str, image: str, *, \n",
    "                 platform: str = 'linux/amd64') -> None:\n",
    "        \"\"\"Interface\n",
    "\n",
    "        Args:\n",
    "            data_folder (str): folder that holds the sample raw data\n",
    "            ref_folder (str): folder that holds all reference files\n",
    "            output_folder (str): folder to hold TRUST4 generated output files\n",
    "            image (str): docker image name\n",
    "            platform (str, optional): image platform. Defaults to 'linux/amd64'.\n",
    "        \"\"\"\n",
    "        self.image = image\n",
    "        self.bind_info: dict[str, str] = {'/data': data_folder, '/ref': ref_folder, '/out': output_folder}\n",
    "        self.base_command: str = f\"docker run --platform {platform} \" + ' '.join(f\"-v {v}:{k}\" for k, v in self.bind_info.items()) + \\\n",
    "        f\" {self.image} run-trust4\"\n",
    "\n",
    "    def usage(self) -> CommandOut:\n",
    "        \"\"\"Shows usage.\"\"\"\n",
    "        return run_command(self.base_command, check=False)\n",
    "\n",
    "    def _generate_command(self, ref_file: str, coordinate_file: str,\n",
    "                          extra_params: str | None = None) -> str:\n",
    "        \"\"\"Command with --ref and -f options.\"\"\"\n",
    "        command = self.base_command + f\" --ref /ref/{ref_file} -f /ref/{coordinate_file} --od /out\"\n",
    "        if extra_params is not None:\n",
    "            command += f\" {extra_params}\"\n",
    "        return command\n",
    "\n",
    "    def generate_bam_command(self, ref_file: str, coordinate_file: str, bam_file: str, \n",
    "                             extra_params: str | None = None) -> str:\n",
    "        \"\"\"Generates the command to analyze a BAM file.\"\"\"\n",
    "        return self._generate_command(ref_file, coordinate_file, extra_params=extra_params) + f\" -b /data/{bam_file}\"\n",
    "    \n",
    "    def generate_paired_fastaq_command(self, ref_file: str, coordinate_file: str, fasta_file1: str, fasta_file2: str, \n",
    "                                       extra_params: str | None = None) -> str:\n",
    "        \"\"\"Generates a command to analyze paired fastaq files.\"\"\"\n",
    "        return self._generate_command(ref_file, coordinate_file, extra_params=extra_params) + f\" -1 /data/{fasta_file1} -2 /data/{fasta_file2}\"\n",
    "    \n",
    "    def generate_single_fastaq_command(self, ref_file: str, coordinate_file: str, fasta_file: str,\n",
    "                                       extra_params: str | None = None) -> str:\n",
    "        \"\"\"Generates a command to analyze single fastaq file.\"\"\"\n",
    "        return self._generate_command(ref_file, coordinate_file, extra_params=extra_params) + f\" -u /data/{fasta_file}\"\n",
    "\n",
    "    def run_bam(self, ref_file: str, coordinate_file: str, bam_file: str,\n",
    "                extra_params: str | None = None) -> CommandOut:\n",
    "        \"\"\"Runs TRUST4 on a BAM input file.\"\"\"\n",
    "        return run_command(\n",
    "            self.generate_bam_command(ref_file, coordinate_file, bam_file, extra_params=extra_params)\n",
    "        )\n",
    "\n",
    "    def run_paired_fastaq(self, ref_file: str, coordinate_file: str, fasta_file1: str, fasta_file2: str,\n",
    "                          extra_params: str | None = None) -> CommandOut:\n",
    "        \"\"\"Runs TRUST4 on paired fastaq files.\"\"\"\n",
    "        return run_command(\n",
    "            self.generate_paired_fastaq_command(ref_file, coordinate_file, fasta_file1, fasta_file2, extra_params=extra_params)\n",
    "        )\n",
    "    \n",
    "    def run_single_fastaq(self, ref_file: str, coordinate_file: str, fasta_file: str,\n",
    "                          extra_params: str | None = None) -> CommandOut:\n",
    "        \"\"\"Runs TRUST4 on a fastaq file.\"\"\"\n",
    "        return run_command(\n",
    "            self.generate_single_fastaq_command(ref_file, coordinate_file, fasta_file, extra_params=extra_params)\n",
    "        )\n",
    "    \n",
    "    def __repr__(self) -> str:\n",
    "        return f\"TRUSTContainer('{self.bind_info['/data']}', '{self.bind_info['/ref']}', '{self.bind_info['/out']}', '{self.image}')\"\n",
    "\n",
    "    def __str__(self) -> str:\n",
    "        return \"TRUST4 docker container interface.\\n\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"Parameters\"></a>\n",
    "## Parameters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# folder to hold the analysis figures \n",
    "figure_path: Path = Path('./Figures')\n",
    "\n",
    "# folder to hold all reference files\n",
    "reference_path: Path = Path('./Reference')\n",
    "\n",
    "# reference files\n",
    "genecode_file: Path = reference_path / 'genecode_annotation.gtf'  # annotation file\n",
    "hg38_ref_file: Path = reference_path / 'Homo_sapients_GRCh38.fa'  # hg38 reference genome fasta file\n",
    "VDJC_gene_ref_file: Path = reference_path / 'IMGT+C.fa'  # reference database sequence (for the --ref file)\n",
    "\n",
    "# BCR and TCR sequences and names\n",
    "bcr_tcr_gene_name_file: Path = reference_path / 'bcr_tcr_gene_name.txt'  # BCR and TCR gene names (required for generating the -f file)\n",
    "bcr_tcr_gene_seq_file: Path = reference_path / 'bcrtcr.fa'  # genomic sequence and coordinate of V,J,C genes (-f option (when using BAM files))\n",
    "\n",
    "# TRUST4 docker\n",
    "data_path: str = os.path.abspath('./Data')  # BAM/fastaq files folders\n",
    "ref_path: str = os.path.abspath(reference_path)  # reference files folder\n",
    "out_path: str = os.path.abspath('./Output')  # output files folder\n",
    "docker_container: str = 'quay.io/biocontainers/trust4:1.1.5--h5ca1c30_0'\n",
    "\n",
    "# SRA toolkit folder\n",
    "sra_command_path: str = './Tools/sratoolkit.3.2.1/bin'\n",
    "\n",
    "# samtools execution command\n",
    "samtools_path: str = str(shutil.which('samtools'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"Configurations\"></a>\n",
    "## Configurations\n",
    "Run this section once to load and generate configurations files and load the TRUST4 docker image."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download genecode annotation file\n",
    "command = f\"wget -qO- https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_47/gencode.v47.annotation.gtf.gz | gunzip > {genecode_file}\"\n",
    "ret, stdout, stderr = run_command(command)\n",
    "print(ret)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download Ensembl GRCh38 genome\n",
    "command = f\"wget -qO- https://ftp.ensembl.org/pub/release-113/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz | gunzip > \\\n",
    "    {hg38_ref_file}\"\n",
    "ret, stdout, stderr = run_command(command)\n",
    "print(ret)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the --ref file\n",
    "command = f\"perl BuildImgtAnnot.pl Homo_sapien > {VDJC_gene_ref_file}\"\n",
    "ret, stdout, stderr = run_command(command)\n",
    "print(ret)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate BCR and TCR gene names (needed for generating the -f file)\n",
    "command = f\"\"\"grep \">\" {VDJC_gene_ref_file} | cut -f2 -d'>' | cut -f1 -d'*' | sort | uniq > {bcr_tcr_gene_name_file}\"\"\"\n",
    "ret, stdout, stderr = run_command(command)\n",
    "print(ret)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the BCR and TCR sequences (for the -f option)\n",
    "command = f\"perl BuildDatabaseFa.pl {hg38_ref_file} {genecode_file} {bcr_tcr_gene_name_file} > {bcr_tcr_gene_seq_file}\"\n",
    "ret, stdout, stderr = run_command(command)\n",
    "print(ret)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the TRUST4 docker image. Make sure docker is installed.\n",
    "command = f\"docker pull {docker_container}\"\n",
    "ret, stdout, stderr = run_command(command)\n",
    "print(ret)\n",
    "\n",
    "# verify\n",
    "!docker image ls | grep {docker_container.split(':')[0]}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Verify (Optional)\n",
    "We run `TRUST4` on the provided example to make sure TRUST4 is running properly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# first show TRUST4 usage\n",
    "example_data_path: str = os.path.abspath('./example')\n",
    "example_ref_path: str = os.path.abspath('./')\n",
    "example_out_path: str = os.path.abspath('./example/Output')\n",
    "# -----------------------------------\n",
    "trust4c_exp = TRUSTContainer(example_data_path, example_ref_path, example_out_path, docker_container)\n",
    "ret, stdout, stderr = trust4c_exp.usage()\n",
    "print(stdout, stderr, sep='\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now run on the example BAM file (with provided reference files)\n",
    "bam_file: Path = Path('example.bam')\n",
    "# -----------------------------------\n",
    "ret, stdout, stderr = trust4c_exp.run_bam('human_IMGT+C.fa', 'hg38_bcrtcr.fa', bam_file.name)\n",
    "print(ret, stdout, stderr, sep='\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# verify\n",
    "!bash mytrust_example_test.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now use the provided paired fastaq files\n",
    "ret, stdout, stderr = trust4c_exp.run_paired_fastaq('human_IMGT+C.fa', 'hg38_bcrtcr.fa', 'example_1.fq', 'example_2.fq')\n",
    "print(ret, stdout, stderr, sep='\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# verify\n",
    "!bash mytrust_example1_test.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# show generated report\n",
    "report_file: Path = Path(example_out_path) / f'TRUST_{bam_file.stem}_report.tsv'\n",
    "df = pd.read_csv(report_file, sep='\\t').rename(columns={'#count': 'count'})\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now run with generated reference files\n",
    "trust4c = TRUSTContainer(data_path, ref_path, out_path, docker_container)\n",
    "ret, stdout, stderr = trust4c.run_bam(VDJC_gene_ref_file.name, bcr_tcr_gene_seq_file.name, bam_file.name)\n",
    "print(ret, stdout, stderr, sep='\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# show example generated report\n",
    "report_file: Path = Path(out_path) / f'TRUST_{bam_file.stem}_report.tsv'\n",
    "df = pd.read_csv(report_file, sep='\\t').rename(columns={'#count': 'count'})\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"Pipeline\"></a>\n",
    "## Pipeline\n",
    "\n",
    "In this section we run TRUST4 on samples.\n",
    "\n",
    "### Sample Data\n",
    "In the example below, we use several samples from the GEO dataset: [The tumor immune microenvironment and therapeutic efficacy of trastuzumab deruxtecan in gastric cancer](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE262648).\n",
    "\n",
    "**Samples choosen**:\n",
    "- SRX24082037: GSM8172959: Sbj3_BL; Homo sapiens; RNA-Seq. Run accession number is [SRR28479475](https://www.ncbi.nlm.nih.gov/sra?term=SRX24082037)\n",
    "- SRX24082038: GSM8172960: Sbj4_BL; Homo sapiens; RNA-Seq. Run accession number is [SRR28479474](https://www.ncbi.nlm.nih.gov/sra?term=SRX24082038)\n",
    "- SRX24082039: GSM8172961: Sbj7_BL; Homo sapiens; RNA-Seq. Run accession number is [SRR28479473](https://www.ncbi.nlm.nih.gov/sra?term=SRX24082039)\n",
    "\n",
    "**Downloading a GEO sample**:\n",
    "- See the `Miscellaneous` section below.\n",
    "\n",
    "[Back to Top](#table-of-contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# instantiate a TRUST4 class\n",
    "trust4c = TRUSTContainer(data_path, ref_path, out_path, docker_container)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' \n",
    "Run the pipeline on a single-end fastaq file.\n",
    "\n",
    "In case of a BAM file, execute instead: \n",
    "    ret, stdout, stderr = trust4c.run_bam(VDJC_gene_ref_file.name, bcr_tcr_gene_seq_file.name, bam_file, extra_params=f\" -t {num_threads}\")\n",
    "\n",
    "In case of paired fasta/fastq files, execute instead:\n",
    "    ret, stdout, stderr = trust4c.run_paired_fastaq(VDJC_gene_ref_file.name, bcr_tcr_gene_seq_file.name, fastaq_file1, fastaq_file2, \n",
    "                                                    extra_params=f\" -t {num_threads}\")\n",
    "\n",
    "Note:\n",
    "- For fastaq input file(s), the coordinate_file input (second input in the run method) can be identical \n",
    "  to the ref_file input (first input in the run method) as the coordinates are not needed in this case. Thus, e.g., running \n",
    "  trust4c.run_single_fastaq(VDJC_gene_ref_file.name, bcr_tcr_gene_seq_file.name, fastaq_file, extra_params=f\" -t {num_threads}\"), or \n",
    "  trust4c.run_single_fastaq(VDJC_gene_ref_file.name, VDJC_gene_ref_file.name, fastaq_file, extra_params=f\" -t {num_threads}\") \n",
    "  will yield the same results.\n",
    "'''\n",
    "#fastaq_file: str = 'SRR28479475.fastq'\n",
    "#fastaq_file: str = 'SRR28479474.fastq'\n",
    "fastaq_file: str = 'SRR28479473.fastq'\n",
    "\n",
    "num_threads: int = 8  # number of required threads\n",
    "# ------------------------------------------------\n",
    "ret, stdout, stderr = trust4c.run_single_fastaq(VDJC_gene_ref_file.name, bcr_tcr_gene_seq_file.name, fastaq_file, extra_params=f\" -t {num_threads}\")\n",
    "print(ret, stdout, stderr, sep='\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# show generated report\n",
    "report_file: Path = Path(out_path) / f'TRUST_{Path(fastaq_file).stem}_report.tsv'\n",
    "df = pd.read_csv(report_file, sep='\\t').rename(columns={'#count': 'count'})\n",
    "print(f\"Report from {report_file}:\")\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"resulteval\"></a>\n",
    "## Result Evaluations\n",
    "\n",
    "In this section we analyze the TRUST4 results.\n",
    "\n",
    "[Back to Top](#table-of-contents)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "VDJC_genes: list[str] = ['V', 'D', 'J', 'C']\n",
    "CDR3_out_cols: list[str] = [\n",
    "    'consensus_id',\n",
    "    'index_within_consensus',\n",
    "    'V_gene',\n",
    "    'D_gene',\n",
    "    'J_gene',\n",
    "    'C_gene',\n",
    "    'CDR1',\n",
    "    'CDR2',\n",
    "    'CDR3', \n",
    "    'CDR3_score', \n",
    "    'read_fragment_count',\n",
    "    'CDR3_germline_similarity',\n",
    "    'complete_vdj_assembly'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set the sample to analyze\n",
    "sample_name: str = 'SRR28479475'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Report Analysis\n",
    "\n",
    "- `TRUST_sample_report.tsv` is a report file focusing on CDR3. It lists unique assembled immune receptor sequence\n",
    "- Consolidates identical CDR3 sequences from `TRUST_cdr3.out` (a contig-driven CDR3 file) and filters out partial or low-confidence sequences\n",
    "- excludes partial CDR3 sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' \n",
    "Note:\n",
    "- the frequency is computed as (read count of a specific TCR [BCR] CDR3 seq) / (total read count of all BCR [TCR] CDR3 seq). It\n",
    "  provides imformation into the relative abundance of each clonotype within the immune repertoire\n",
    "'''\n",
    "figsize: tuple[int, int] = (7,4)\n",
    "save_figure: bool = False\n",
    "# -------------------------------\n",
    "report_file: Path = Path(out_path) / f'TRUST_{sample_name}_report.tsv'\n",
    "df = pd.read_csv(report_file, sep='\\t').rename(columns={'#count': 'count'})\n",
    "print(f\"Report for {sample_name}:\")\n",
    "display(df)\n",
    "\n",
    "print(f\"- {df['cid'].unique().size:,} unique assemblies\")\n",
    "\n",
    "# number of CDR3 sequences found\n",
    "num_nt_cdr3: int = df['CDR3nt'].unique().size\n",
    "num_aa_cdr3: int = df['CDR3aa'].unique().size\n",
    "print(f\"- {num_nt_cdr3:,} ({num_aa_cdr3:,}) unique CDR3 NT (AA) sequences\")\n",
    "\n",
    "# complete VDJ variable domain\n",
    "cid_full_length: pd.Series = df['cid_full_length']\n",
    "print(f\"- {100 * cid_full_length.sum() / cid_full_length.size:.1f}% of {cid_full_length.size:,} receptors contain complete variable domain (V(D)J)\")\n",
    "\n",
    "# read_counts and frequency\n",
    "read_counts: pd.Series = df['count']\n",
    "frequency: list[float] = list(100 * df['frequency'])\n",
    "fig, ax = plt.subplots(nrows=2, ncols=1, tight_layout=True, figsize=figsize, squeeze=True)\n",
    "ax[0].plot(read_counts)\n",
    "ax[0].grid()\n",
    "ax[0].set_ylabel('count')\n",
    "\n",
    "ax[1].plot(frequency)\n",
    "ax[1].grid()\n",
    "ax[1].set_ylabel('freq [%]')\n",
    "ax[1].set_xlabel('receptor index')\n",
    "fig.suptitle(f'{sample_name}: CDR3 read count and frequency per receptor', fontsize=12)\n",
    "if save_figure:\n",
    "    fig_file: Path = figure_path / f\"{sample_name}_CDR3_read_counts_and_freq.jpg\"\n",
    "    plt.savefig(fig_file)\n",
    "    print(f\"\\nFigure saved in {fig_file}.\")\n",
    "\n",
    "# read count stats\n",
    "stats = read_counts.describe().drop('count')\n",
    "#display(stats)\n",
    "stats_verbose = {x: round(v, 1) for x, v in stats.items()}\n",
    "print(f\"- CDR3 read count stats: {stats_verbose}\")\n",
    "fig, ax = plt.subplots(nrows=1, ncols=1, tight_layout=True, figsize=figsize, squeeze=True)\n",
    "ax.plot(stats.index, list(stats.values), '--*', linewidth=1.0, markersize=12)\n",
    "ax.set_title(f\"{sample_name}: CDR3 read count descriptive statistics\", fontsize=12)\n",
    "ax.yaxis.grid()\n",
    "if save_figure:\n",
    "    fig_file: Path = figure_path / f\"{sample_name}_CDR3_read_counts_stats.jpg\"\n",
    "    plt.savefig(fig_file)\n",
    "    print(f\"\\nFigure saved in {fig_file}.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clonotype \n",
    "Clonotype for TCR are the CDR3 sequence and for BCR are the cluster with the same V, J gene assignments and similar CDR3 sequences\n",
    "\n",
    "We note that:\n",
    "- most are BCR\n",
    "- for a given V and J gene set, number of CDR3 sequence is in general low (mostly 1 or 2)\n",
    "- can use `Bio.pairwise2` to cluster further within a V and J gene set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figsize: tuple[int, int] = (7,4)\n",
    "save_figure: bool = False\n",
    "# -------------------------------\n",
    "v_genes = df['V'].unique().tolist()\n",
    "v_genes_tcr = [v for v in v_genes if v.startswith('T')]\n",
    "v_genes_bcr = [v for v in v_genes if v.startswith('I')]\n",
    "print(f\"BCR: {len(v_genes_bcr)}, TCR: {len(v_genes_tcr)}\")\n",
    "\n",
    "# df_tcr = df.query(f\"V in {v_genes_tcr}\")\n",
    "df_bcr = df.query(f\"V in {v_genes_bcr}\")\n",
    "\n",
    "# lower bound estimate\n",
    "df_clonotype = (df_bcr[['V', 'J']]\n",
    "                .query(\"V != '.' and J != '.'\")\n",
    "                .groupby(by=['V', 'J'])\n",
    "                .size()\n",
    "                .to_frame(name='size').sort_values(by='size', ascending=False)\n",
    "                )\n",
    "print(f\"\\n- number of BCR clonotypes >= {df_clonotype.shape[0]:,}\")\n",
    "display(df_clonotype)\n",
    "\n",
    "fig, ax = plt.subplots(nrows=1, ncols=1, tight_layout=True, figsize=figsize, squeeze=True)\n",
    "ax.plot(df_clonotype.values)\n",
    "ax.grid()\n",
    "ax.set_xlabel('clonotype index')\n",
    "ax.set_ylabel('size')\n",
    "ax.set_title(f\"{sample_name}: BCR clonotype diversity (lower estimation)\", fontsize=12)\n",
    "if save_figure:\n",
    "    fig_file: Path = figure_path / f\"{sample_name}_clonotype_diversity.jpg\"\n",
    "    plt.savefig(fig_file)\n",
    "    print(f\"\\nFigure saved in {fig_file}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VDJC genes diversity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figsize: tuple[int, int] = (7,4)\n",
    "width: float = 0.6\n",
    "save_figure: bool = False\n",
    "# -------------------------------\n",
    "VDJC_genes_diverisity = defaultdict(list)\n",
    "for gene in VDJC_genes:\n",
    "    genes = df[gene].unique().tolist()\n",
    "    tcr_genes = [v for v in genes if v.startswith('T')]\n",
    "    bcr_genes = [v for v in genes if v.startswith('I')]\n",
    "    print(f\"- {gene} gene: total of {len(genes)} genes: {len(bcr_genes)} BCR genes and {len(tcr_genes)} TCR genes\")\n",
    "    VDJC_genes_diverisity['BCR'].append(len(bcr_genes))\n",
    "    VDJC_genes_diverisity['TCR'].append(len(tcr_genes))\n",
    "\n",
    "# plot\n",
    "fig, ax = plt.subplots(nrows=1, ncols=1, tight_layout=True, figsize=figsize, squeeze=True)\n",
    "bottom = np.zeros(len(VDJC_genes), dtype=int)\n",
    "for gene, gene_bcr_tcr_count in VDJC_genes_diverisity.items():\n",
    "    ax.bar(VDJC_genes, gene_bcr_tcr_count, width, label=gene, bottom=bottom)\n",
    "    bottom += gene_bcr_tcr_count\n",
    "ax.grid()\n",
    "ax.set_title(f\"{sample_name}: Diversity of {','.join(VDJC_genes)} genes\", fontsize=12)\n",
    "ax.legend()\n",
    "\n",
    "if save_figure:\n",
    "    fig_file: Path = figure_path / f\"{sample_name}_VDJC_diversity.jpg\"\n",
    "    plt.savefig(fig_file)\n",
    "    print(f\"\\nFigure saved in {fig_file}.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CDRs Information\n",
    "\n",
    "- `TRUST_sample_cdr3.out` reports the CDR1,2,3 and gene information for each consensus assemblies\n",
    "- contains both complete and partial CDR3 sequences, which are likely to be false positive or partially expressed V, J genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' \n",
    "Note:\n",
    "- index_within_consensus: due to multiple CDR3s within a single consensus sequence.\n",
    "'''\n",
    "figsize: tuple[int, int] = (7,4)\n",
    "save_figure: bool = False\n",
    "# --------------------------------\n",
    "cdr3_file: Path = Path(out_path) / f'TRUST_{sample_name}_cdr3.out'\n",
    "df_cdr3 = pd.read_csv(cdr3_file, sep='\\t', names=CDR3_out_cols)\n",
    "display(df_cdr3)\n",
    "\n",
    "print(f\"- {df_cdr3['consensus_id'].unique().size:,} unique consensus IDs\")\n",
    "\n",
    "# CDR3 score\n",
    "cdr3_scores = list(df_cdr3['CDR3_score'])\n",
    "# plot\n",
    "fig, ax = plt.subplots(nrows=1, ncols=1, tight_layout=True, figsize=figsize, squeeze=True)\n",
    "ax.hist(cdr3_scores, bins=list(np.arange(0, 1, 0.05, dtype=float)))\n",
    "ax.grid()\n",
    "ax.set_title(f\"{sample_name}: CDR3 score distribution over {len(cdr3_scores):,} consensus assemblies\", fontsize=12)\n",
    "\n",
    "if save_figure:\n",
    "    fig_file: Path = figure_path / f\"{sample_name}_CDR3_scores.jpg\"\n",
    "    plt.savefig(fig_file)\n",
    "    print(f\"\\nFigure saved in {fig_file}.\")\n",
    "#print(\"Recall that 0 score implies partial CDR3.\")\n",
    "\n",
    "# somatic hypermutation\n",
    "CDR3_germline_similarity: list[float] = df_cdr3['CDR3_germline_similarity'].tolist()\n",
    "fig, ax = plt.subplots(nrows=1, ncols=1, tight_layout=True, figsize=figsize, squeeze=True)\n",
    "ax.hist(CDR3_germline_similarity, bins=list(np.arange(0, 100, 5, dtype=float)))\n",
    "ax.grid()\n",
    "ax.set_title(f\"{sample_name}: germline similarity (est. of somatic hypermutation)\", fontsize=12)\n",
    "ax.set_xlabel('%')\n",
    "\n",
    "if save_figure:\n",
    "    fig_file: Path = figure_path / f\"{sample_name}_germline_similarity.jpg\"\n",
    "    plt.savefig(fig_file)\n",
    "    print(f\"\\nFigure saved in {fig_file}.\")\n",
    "\n",
    "# CDR3 sequence length distributions\n",
    "# Note: CDR3 sequence varies in each assembly\n",
    "CDRs_seq_cols: list[str] = ['CDR1', 'CDR2', 'CDR3']\n",
    "for cdr_col in CDRs_seq_cols:\n",
    "    cdr_seqs = df_cdr3[cdr_col]\n",
    "    cdr_seq_lens = cdr_seqs.apply(len)\n",
    "    cdr_seq_lens = cdr_seq_lens[cdr_seq_lens>1]\n",
    "    stats = cdr_seq_lens.describe().drop('count')\n",
    "    stats_verbose = {x: round(v, 1) for x, v in stats.items()}\n",
    "    print(f\"- {cdr_seqs.unique().size:,} unique {cdr_col} sequences.\\tseq len stats: {stats_verbose}\")\n",
    "\n",
    "# genes usage\n",
    "for gene in [f\"{g}_gene\" for g in VDJC_genes]:\n",
    "    print(f\"- {df_cdr3[gene].unique().size:,} unique {gene} are used\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"resultsum\"></a>\n",
    "## Results Summary\n",
    "\n",
    "**<u>SRR28479473</u>** (19.9M spots, 1.5G bases)\n",
    "- 450 unique assemblies\n",
    "- 638 (525) unique CDR3 NT (AA) sequences\n",
    "- 4.0% of 642 receptors contain complete variable domain (V(D)J)\n",
    "- CDR3 read count stats: {'mean': 4.2, 'std': 10.3, 'min': 1.0, '25%': 1.0, '50%': 2.0, '75%': 3.0, 'max': 125.0}\n",
    "- number of BCR clonotypes >= 174\n",
    "- V gene: total of 80 genes: 78 BCR genes and 1 TCR genes\n",
    "- D gene: total of 27 genes: 26 BCR genes and 0 TCR genes\n",
    "- J gene: total of 23 genes: 21 BCR genes and 1 TCR genes\n",
    "- C gene: total of 9 genes: 7 BCR genes and 1 TCR genes\n",
    "- 17 unique CDR1 sequences.\tseq len stats: {'mean': 25.4, 'std': 6.9, 'min': 18.0, '25%': 18.0, '50%': 24.0, '75%': 33.0, 'max': 36.0}\n",
    "- 38 unique CDR2 sequences.\tseq len stats: {'mean': 13.8, 'std': 7.0, 'min': 8.0, '25%': 9.0, '50%': 9.0, '75%': 24.0, 'max': 30.0}\n",
    "\n",
    "\n",
    "**<u>SRR28479474</u>** (37.3M spots, 2.8G bases)\n",
    "- 232 unique assemblies\n",
    "- 335 (272) unique CDR3 NT (AA) sequences\n",
    "- 0.0% of 340 receptors contain complete variable domain (V(D)J)\n",
    "- CDR3 read count stats: {'mean': 5.2, 'std': 8.2, 'min': 1.0, '25%': 1.0, '50%': 2.0, '75%': 5.0, 'max': 58.0}\n",
    "- number of BCR clonotypes >= 109\n",
    "- V gene: total of 73 genes: 63 BCR genes and 9 TCR genes\n",
    "- D gene: total of 17 genes: 15 BCR genes and 1 TCR genes\n",
    "- J gene: total of 25 genes: 16 BCR genes and 8 TCR genes\n",
    "- C gene: total of 7 genes: 4 BCR genes and 2 TCR genes\n",
    "- 6 unique CDR1 sequences.\tseq len stats: {'mean': 22.1, 'std': 5.7, 'min': 18.0, '25%': 18.0, '50%': 18.0, '75%': 28.5, 'max': 30.0}\n",
    "- 17 unique CDR2 sequences.\tseq len stats: {'mean': 12.6, 'std': 6.1, 'min': 9.0, '25%': 9.0, '50%': 9.0, '75%': 18.0, 'max': 24.0}\n",
    "\n",
    "\n",
    "**<u>SRR28479475</u>** (21.3M spots, 1.6G bases)\n",
    "- 701 unique assemblies\n",
    "- 1,015 (826) unique CDR3 NT (AA) sequences\n",
    "- 3.8% of 1,026 receptors contain complete variable domain (V(D)J)\n",
    "- CDR3 read count stats: {'mean': 3.2, 'std': 7.3, 'min': 1.0, '25%': 1.0, '50%': 1.0, '75%': 3.0, 'max': 103.0}\n",
    "- number of BCR clonotypes >= 241\n",
    "- V gene: total of 119 genes: 113 BCR genes and 5 TCR genes\n",
    "- D gene: total of 30 genes: 28 BCR genes and 1 TCR genes\n",
    "- J gene: total of 28 genes: 23 BCR genes and 4 TCR genes\n",
    "- C gene: total of 8 genes: 5 BCR genes and 2 TCR genes\n",
    "- 20 unique CDR1 sequences.\tseq len stats: {'mean': 24.5, 'std': 6.7, 'min': 18.0, '25%': 18.0, '50%': 24.0, '75%': 33.0, 'max': 33.0}\n",
    "- 45 unique CDR2 sequences.\tseq len stats: {'mean': 11.5, 'std': 5.6, 'min': 9.0, '25%': 9.0, '50%': 9.0, '75%': 9.0, 'max': 24.0}\n",
    "\n",
    "**<u>main (and preliminary) observations</u>**\n",
    "- note that `SRR28479474` > `SRR28479475` > `SRR28479473` in terms of total sequenced size\n",
    "- consistency in number of sequences, genes, and clonoytpe diversity ranking among samples (`SRR28479474` < `SRR28479473` < `SRR28479475`), somewhat opposite to total sequenced size (?)\n",
    "- CDR3 scores and germline similarity (measure of somatic hypermutations) ranking consistent among samples (see the `_CDR3_scores,jpg` figures)\n",
    "- similar (and low) CDR3 mean read counts among samples\n",
    "- similar D, J, C gene usage among samples\n",
    "- V gene usage consistant with number of CDR3 sequences among samples\n",
    "- `SRR28479474` contains more receptors with relatively high read count frequency (proportion of read_count mapped to specific BCR/TCR), than the other samples (see the `_CDR3_read_counts_and_freq,jpg` figures). Note that this sample ranked the lowest in terms of sequences and gene found but highest in terms of total sequenced size\n",
    "\n",
    "\n",
    "[Back to Top](#table-of-contents)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Miscellaneous"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"donaloaddata\"></a>\n",
    "## Download GEO Samples\n",
    "\n",
    "Below we implement a SRAtoolkit interface and use it to download samples.\n",
    "\n",
    "#### Pre-requisite\n",
    "We are using the SRA toolkit to download the GEO samples:\n",
    "- download the [SRA toolkit](https://github.com/ncbi/sra-tools/wiki/01.-Downloading-SRA-Toolkit) (choose the right architecture)\n",
    "- once downloaded, execute (example here is form MAC with ARM CPU) `tar -xzvf sratoolkit.3.2.1-mac-arm64.tar`\n",
    "\n",
    "The SRA toolkit resides in `./Tools/sratoolkit.3.2.1/bin` folder.\n",
    "\n",
    "#### Retrieve a GEO sample ID\n",
    "1. Go to the GEO [Dataset](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE262648)\n",
    "1. Click on a sample from the `Samples` row (e.g., `GSM8172959`)\n",
    "1. Click on the SRA accession number at the bottom (under `Relations`). This will open the corresponding SRA page. Copy the corresponding SRA run accession (e.g. `SRR28479474`)\n",
    "1. Use the retreived SRA run accession below\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This should move to a ./Util/SRAtoolkit_api.py file.\n",
    "\n",
    "class SRAtoolkit():\n",
    "    \"\"\"SRA toolkit Interface.\"\"\"\n",
    "    def __init__(self, sra_accession: str, folder: Path,\n",
    "                 sratoolkit_bin_folder: Path = Path('./Tools/sratoolkit.3.2.1/bin')) -> None:\n",
    "        \"\"\"Interface.\n",
    "\n",
    "        Args:\n",
    "            sra_accession (str): the SRA run accession number (e.g. SRR28479474).\n",
    "            folder (Path): destination folder (will hold the .sra, .bam, abd .fastaq files).\n",
    "            sratoolkit_bin_folder (Path, optional): path to sratoolkit bin/ folder. Defaults to Path('./Tools/sratoolkit.3.2.1/bin').\n",
    "        \"\"\"\n",
    "        self.sra_accession: str = sra_accession\n",
    "        self.folder: Path = folder\n",
    "        self.sratoolkit_bin_folder: Path = sratoolkit_bin_folder\n",
    "        self.sra_file: Path = folder / f\"{sra_accession}.sra\"\n",
    "\n",
    "    def stats(self) -> str:\n",
    "        \"\"\"Returns the stats of the sample in XML formatted string.\"\"\"\n",
    "        command: str = f\"{self.sratoolkit_bin_folder}/sra-stat -x {self.sra_accession}\"\n",
    "        try:\n",
    "            _, stdout, _ = run_command(command)\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print(f\"stats failed: {e}\")\n",
    "            print(f\"Error output: {e.stderr}\")\n",
    "        return stdout\n",
    "    \n",
    "    def info(self) -> str:\n",
    "        \"\"\"Returns the sample information.\"\"\"\n",
    "        command: str = f\"{self.sratoolkit_bin_folder}/vdb-dump --info {self.sra_accession}\"\n",
    "        try:\n",
    "            _, stdout, _ = run_command(command)\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print(f\"info failed: {e}\")\n",
    "            print(f\"Error output: {e.stderr}\")\n",
    "        return stdout\n",
    "\n",
    "    def prefetch(self) -> CommandOut:\n",
    "        \"\"\"Downloads the .sra file.\"\"\"\n",
    "        command: str = f\"{self.sratoolkit_bin_folder}/prefetch -o {self.sra_file} {self.sra_accession}\"\n",
    "        try:\n",
    "            ret, stdout, stderr = run_command(command)\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print(f\"prefetch failed: {e}\")\n",
    "            print(f\"Error output: {e.stderr}\")\n",
    "        return ret, stdout, stderr \n",
    "    \n",
    "    def gen_fastaq(self, temp_folder: Path | None = None) -> CommandOut:\n",
    "        \"\"\"Generates fastaq file(s) from the .sra file.\n",
    "\n",
    "        Args:\n",
    "            temp_folder (Path | None, optional): temporary folder. Defaults to the data folder.\n",
    "\n",
    "        Returns:\n",
    "            CommandOut: command output.\n",
    "        \"\"\"\n",
    "        if temp_folder is None:\n",
    "            temp_folder = self.folder\n",
    "        command: str = f\"{self.sratoolkit_bin_folder}/fasterq-dump {self.sra_file} -t {temp_folder} -O {self.folder}\"\n",
    "        try:\n",
    "            ret, stdout, stderr = run_command(command)\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print(f\"generating fastaq failed: {e}\")\n",
    "            print(f\"Error output: {e.stderr}\")\n",
    "        return ret, stdout, stderr\n",
    "    \n",
    "    def gen_bam(self, samtools_exe: str, remove_sam: bool = True) -> tuple[CommandOut, CommandOut]:\n",
    "        \"\"\"Generates the BAM file from the .sra file. Note that a SAM file is generated first.\n",
    "\n",
    "        Args:\n",
    "            samtools_exe (str): samtool execution command (used to convert the SAM file to BAM file).\n",
    "            remove_sam (bool, optional): remove the SAM file after generating the BAM file. Defaults to True.\n",
    "\n",
    "        Returns:\n",
    "            tuple[CommandOut, CommandOut]: SAM generation and conversion to BAM commands output.\n",
    "        \"\"\"\n",
    "        # gen the sam file\n",
    "        sam_file = self.sra_file.with_suffix('.sam')\n",
    "        command: str = f\"{self.sratoolkit_bin_folder}/sam-dump {self.sra_file} > {sam_file}\"\n",
    "        try:\n",
    "            rets, stdouts, stderrs = run_command(command)\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print(f\"generating sam failed: {e}\")\n",
    "            print(f\"Error output: {e.stderr}\")\n",
    "\n",
    "        # convert to bam\n",
    "        bam_file = sam_file.with_suffix('.bam')\n",
    "        command: str = f\"{samtools_exe} view -bS {sam_file} > {bam_file}\"\n",
    "        try:\n",
    "            retb, stdoutb, stderrb = run_command(command)\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print(f\"generating bam failed: {e}\")\n",
    "            print(f\"Error output: {e.stderr}\")\n",
    "\n",
    "        # remove sam\n",
    "        if remove_sam:\n",
    "            sam_file.unlink()\n",
    "\n",
    "        return (rets, stdouts, stderrs), (retb, stdoutb, stderrb)\n",
    "\n",
    "    def cleanup(self) -> None:\n",
    "        \"\"\"Removes the .sra file.\"\"\"\n",
    "        if self.sra_file.is_file():\n",
    "            self.sra_file.unlink()\n",
    "\n",
    "    def prefetch_bam(self, samtools_exe: str, remove_sam: bool = True) -> None:\n",
    "        \"\"\"Wrapper to prefetch and gen_bam\"\"\"\n",
    "        self.prefetch()\n",
    "        self.gen_bam(samtools_exe, remove_sam=remove_sam)\n",
    "        self.cleanup()\n",
    "\n",
    "    def prefetch_fastaq(self, temp_folder: Path | None = None) -> None:\n",
    "        \"\"\"Wrapper to prefetch and gen_fastaq\"\"\"\n",
    "        self.prefetch()\n",
    "        self.gen_fastaq(temp_folder=temp_folder)\n",
    "        self.cleanup()\n",
    "\n",
    "    def __repr__(self) -> str:\n",
    "        return f\"SRAtoolkit('{self.sra_accession}', '{self.folder}')\"\n",
    "\n",
    "    def __str__(self) -> str:\n",
    "        return \"SRA toolkit interface.\\n\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' \n",
    "Fetch a sample and generate the bam/fastaq files.\n",
    "'''\n",
    "sra_sample_accession: str = 'SRR28479473'  # run accession number\n",
    "# ----------------------------------------------------------------\n",
    "srat = SRAtoolkit(sra_sample_accession, Path(data_path))\n",
    "\n",
    "# another option is to inspect the stats xml file\n",
    "aligned = 'align' in srat.info()\n",
    "print(f\"{sra_sample_accession} contains {'aligned' if aligned else 'un-aligned'} data.\")\n",
    "\n",
    "# prefetch\n",
    "ret, stdout, stderr = srat.prefetch()\n",
    "print(ret, stdout, stderr)\n",
    "\n",
    "if aligned:\n",
    "    # generate bam\n",
    "    (rets, stdouts, stderrs), (retb, stdoutb, stderrb) = srat.gen_bam(samtools_path)\n",
    "    print(rets, stdouts, stderrs, retb, stdoutb, stderrb)\n",
    "else:\n",
    "    # generate fastaq\n",
    "    ret, stdout, stderr = srat.gen_fastaq()\n",
    "    print(ret, stdout, stderr)\n",
    "srat.cleanup()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
